[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104427
* __->__ #104418
* #104409
* #104408
* #104407
* #104346

The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

Commit:
[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

voznesenskym committed:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -417,6 +417,7 @@ def __init__(
                 "forward_prefetch": forward_prefetch,
                 "limit_all_gathers": limit_all_gathers,
                 "use_orig_params": use_orig_params,
+                "ignored_states": self._ignored_params,
```
why do we need to propagate ignored_states, but not ignored_modules?

awgu replied:
Good point. Maybe we should also propagate `ignored_modules`. This might be another bug.

awgu replied:
The reason not propagating `ignored_modules` does not break correctness today is because we _only_ use `ignored_modules` to compute the ignored parameters and to compute which modules to ignore for auto wrapping. 

For those two functionalities, we do not need the nested FSDP instances to have `ignored_modules`. However, we should probably still propagate it in case we use `ignored_modules` for something else in the future.

rohan-varma replied:
I see. In general, what do you think about having both ignored_states and ignored_modules? this seems like it can get confusing to users, shall we just consolidate to ignored_states?

rohan-varma commented on test/distributed/fsdp/test_fsdp_ignored_modules.py:
```diff
@@ -173,29 +165,32 @@ def _test_ignored_modules_transformer(
             deterministic=True,
         )
+        if use_auto_wrap:
+            nonwrapped_model.output_proj.weight = nn.Parameter(
```
why do we need this reassignment?

awgu replied:
See the comment above about unsharing the weight.

awgu replied:
I can duplicate the comment here if it helps.

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

Commit:
Update on "[FSDP][5/N] Unblock `ignored_states` + auto wrap (for now)"


The "for now" is because we still have the issue that when using the parameter `ignored_states` path, we do not recover the ignored modules, so FSDP still wraps those as empty shells (no managed parameters), which is not ideal. This is not a blocking issue as far as I know.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
